### DISCUSSION TITLE
Please update all dependencies before each major release

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#2533] [General]
**Is your feature request related to a problem? Please describe.**

I have filed a number of bug reports and PRs to update different dependencies of Serverpod.

I understand Serverpod currently adopts a conservative approach to package upgrades -- a major upgrade of a dependency is only made if there is a clear problem that it solves. Otherwise, old versions are used for all dependencies, sometimes with a version range, to ensure maximal compatibility.

However, this breaks user code, when a user is trying to pull in a large number of their own dependencies in addition to Serverpod. Serverpod's old dependency versions hold back the ability of a user to use the latest version of whatever packages they want from pub.dev.

**Describe the solution you'd like**

Google's strategy for this is to always release packages with all dependency versions set to the latest possible version at time of release. The philosophy is, if you want to depend on an older version of a transitive dependency of a Google package, then you need to depend on an older version of the Google package.

This puts the onus on the end-user to upgrade their package dependencies, if they want the latest features and bugfixes of Google packages. But it also gives the user the ability to use any version they want of dependencies -- they just may have to use an older version of a Google dependency if they want to add a direct dependency in their project on some older package.

*Serverpod really needs to do the same thing. Every time a major release is pushed out, all dependencies of all Serverpod modules and packages should be upgraded to the latest stable versions.*

**Describe alternatives you've considered**

`dependency_overrides` doesn't solve the problem. I have had to fork packages that Serverpod depends upon if I want to depend upon (and be compatible with) both Serverpod and that package. Then I add my forked version as a dependency override. Before the recent changes to Serverpod's firebase deps, I had to fork something like 5 different GitHub projects to override all the transitive versions of the firebase library that was being used, in order to depend upon newer versions of Google's core libraries. It's a mess, and it's not a great solution, since every end user of Serverpod that wants to use newer dependencies has to jump through hoops like this.

Another example of this is https://github.com/serverpod/serverpod/pull/2426 , which I just filed: I need to use a range of Firebase APIs in my own project, but I have to use old versions because Serverpod uses old versions. To upgrade the Firebase dep versions in Serverpod, I had to upgrade the `dart` dep to 3.3, and that requires an update of `web`, which requires an update of `flutter`. Serverpod depends upon old versions of all of these.

**Comment by SandPod** [comment]
First of all, I understand the frustration around this. As you said, our aim with the versions we set for our dependencies is to be as non-intrusive as possible, mindful of your time as a developer. We aim to have a range that always supports the latest stable version.

When we find ourselves behind, preventing you from using the latest stable releases of packages, this is a serious issue that we want to address as quickly as possible.

But we (for now) are human and have a limited capacity.

PRs that help us update dependencies are greatly appreciated since they can often be validated directly towards our testing framework. A green CI with a dependency bump is likely to be merged.

## Update all dependencies before every major release
YES - that is what we are aiming at. When we don't succeed with this, that is a failure on our part, mostly connected to our limited capacity.

To clarify the ask here, you are thinking about the latest version on dependencies and not the SDK or Flutter version, right?

I'm not really sure how we should complete this issue? Maybe add a clarification around our policy in our Readme?

## Google packages bumping to latest SDK.
This is a major frustration for package developers. We just came back from a package ecosystem summit where this was brought up as a pain point for many package maintaners, and Google are reviewing their processes around this. The problem is that an SDK bump generates issues on all packages

**Comment by lukehutch** [comment]
I think the good first step to take is to write a tool that goes through all pubspecs and checks if the current version range includes the latest stable release. Yes please check Dart and Flutter ranges too.

> The problem is that an SDK bump generates issues on all packages

Yes, but I would argue that the problem is much worse the way things are right now. In order to circumvent an old transitive dependency version, every single app developer has to implement their own painful workaround.

**Comment by SandPod** [comment]
That makes sense, DoD for this issue is adding a step in out automated scripts that we run before creating a new Serverpod version.

**Comment by lukehutch** [comment]
A bit more perspective: the two extremes are either: 

1. hold back all your users on old (potentially broken) versions of everything, preventing them from making use of a lot of newer versions of other libraries on pub.dev; or
2. force your users to upgrade all their dependencies to work with the latest versions that Serverpod depends upon.

The only pros of 1 are that it reduces the work the Serverpod team has to do. There are a lot of cons for users, as I discussed previously, including end user frustration, and having to hack together solutions, like forking packages and depending on your own fork in git, to solve an annoying bug that is fixed in a newer version of a dep, but which you can't depend upon because Serverpod depends on the broken older version.

The pros of 2 are manifold:
* It allows users to depend on older versions of many pub.dev libraries if they don't want to upgrade the other dependencies, by simply sticking to an older version of Serverpod. The only library they don't get the latest features of is Serverpod itself. This can dramatically reduce the number of affected libraries that have to be forked to work around problems with old versions of deps.
* Alternatively users can choose to upgrade all their dependency versions, and depend upon the latest version of Serverpod too. Now all the freedom and flexibility as far as which dependency versions a user wants to depend upon is back in the hands of the user.
* Users can know that Serverpod depends on packages with all the latest bugfixes.
* The Serverpod team will have done the work to upgrade to the latest dep versions at each release, which prevents bitrot.

The only con of 2 is that you have to do more work on your end, but you prevent users having to do what is potentially an enormous amount of work (which is multiplied by a potentially unlimited number of users). I probably spent at least 10 hours hacking together a fix so I could depend upon Serverpod when it depended upon a broken version of the unmaintained Firebase library.

I simply don't see any reason to go with 1.

**Comment by SandPod** [comment]
@lukehutch We agree with you.

As said:

> We aim to have a range that always supports the latest stable version.

and

> When we don't succeed with this, that is a failure on our part, mostly connected to our limited capacity.

We are not aiming to go with your first suggestion.
Supporting a range doesn't contradict this statement. If possible, we will always aim to support the previous version as long as it does not prevent us from supporting the latest stable version.

**Comment by lukehutch** [comment]
OK, that is reasonable. As long as the latest version is always supported in the range at each new release. Thanks.

**Comment by vlidholt** [comment]
I'm converting this to a discussion instead. Not really an issue we can take immediate action on.
